MPP-3352: Add first Glean metrics to measure email mask usage

jwhitlock opened the pull request:
Add Glean metrics for measuring P0 email mask usage. This includes email mask creation, label updates, and deletion, as well as receiving emails (forwarded or blocked) through those masks.

This is tracked as epic MPP-3352. The related bug for data review is [bug 1882565](https://bugzilla.mozilla.org/show_bug.cgi?id=1882565).

Some parts of this PR:

* Add the requirements `glean_parser`, to generate Python code from Glean YAML, and `django-ipware`, to extract the user's IP address from headers
* Add a Glean metrics YAML file at `telemetry/glean/relay-server-metrics.yaml`. In the future, this folder can hold front-end or other application metrics, and specification for other telemetry methods.
* Add the generated Python code at `privaterelay/glean/server_events.py`
* Add a CircleCI job to confirm that the generated Python code has not been modified, as well as code to regenerate the file after a requirement or metrics update.
* Add `privaterelay/glean_interface.py` to simplify data extraction and opt-out checking, and to publish events as log entries formatted with MozLog. Add tests for the emitted logs.
* Add `Profile` properties to determine the Relay user's subscription from current data. In the future, we may query Mozilla Accounts for more detailed information on the user's plan term (monthly or yearly).
* Add glean event hooks to the API and email processing. Add tests for the emitted logs.

## Glean Data

This adds five Glean events that have been identified as critical for creating a user engagement model for the email mask feature:

* `email_mask.created` - An email mask is created
* `email_mask.label_updated` - An email mask's label is updated
* `email_mask.deleted` - An email mask is deleted
* `email.forwarded` - An email to or from an email mask is forwarded
* `email.blocked` - An email to or from an email mask is not forwarded

Some data is sent with each event type, identified with a `*` below. Others are specific to event types.

| Event | Measurement Name | Measurement Description |
| :-- | :-- | :-- |
|`*`|`#client_id`| Firefox client ID|
|`*`|`#fxa_id` | Mozilla accounts user ID |
|`*`|`#platform` | Relay client platform |
|`*`|`#n_random_masks` | Number of random masks | 
|`*`|`#n_domain_masks` | Number of premium subdomain masks |
|`*`|`#n_deleted_random_masks` | Number of deleted random masks |
|`*`|`#n_deleted_domain_masks` | Number of deleted domain masks |
|`*`|`#date_joined_relay` | Timestamp for joining Relay, seconds since epoch |
|`*`|`#premium_status` | Subscription type and term |
|`*`|`#date_joined_premium` | Timestamp for starting `premium_status` subscription, seconds since epoch, -1 if not subscribed |
|`*`|`#has_extension` | The user has the Relay Add-on |
|`*`|`#date_got_extension` | Timestamp for adding Relay Add-on, seconds since epoch, -1 if not used |
|`*`|`#mask_id` | Mask ID, 'R' or 'D' followed by a number |
|`*`|`#is_random_mask` | The mask is a random mask, instead of a domain mask |
|`email_mask.created`|`#created_by_api` | The mask was created via the API, rather than an incoming email |
|`email_mask.created`|`#has_website` | The mask was created by the Add-on or integration on a website|
|`email.*`|`#is_reply` | The email is a reply from the Relay user |
|`email.blocked`|`#reason` | Code describing why the email was blocked |
|`email.blocked`|`#can_retry` | The email processes can retry the email, so it may fail multiple times |

## How to test:

* Install the requirements
* Run the script to regenerate the Python interface:
  `bash .circleci/python_job.bash run build_glean` 
    - [ ] There are no errors. `git` does not show any changed files
* Run the script to check the Python generated code:
  `bash .circleci/python_job.bash run check_glean`
    - [ ] There are no errors. Final output is `✓ Files are identical`
* Run the server. Log in as a user with Mozilla accounts telemetry turned on (the default). Create a mask.
    - [ ] Among others, there should be a `glean-server-event` log,  followed by a `request.summary` log, entry like:
    ```json
    {"Timestamp": 1709148528194074112, "Type": "glean-server-event", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.local", "EnvVersion": "2.0", "Severity": 6, "Pid": 87746, "Fields": {"document_namespace": "relay-backend", "document_type": "events", "document_version": "1", "document_id": "695328d4-24fb-42b3-bd4d-853e92efbc0d", "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0", "ip_address": "", "payload": "{\"metrics\": {}, \"events\": [{\"category\": \"email_mask\", \"name\": \"created\", \"extra\": {\"client_id\": \"\", \"fxa_id\": \"fac5c63eb1c04dd4b9b0192f1a5483cb\", \"platform\": \"\", \"n_random_masks\": \"4\", \"n_domain_masks\": \"0\", \"n_deleted_random_masks\": \"1\", \"n_deleted_domain_masks\": \"0\", \"date_joined_relay\": \"1709058088\", \"premium_status\": \"email_unknown\", \"date_joined_premium\": \"-1\", \"has_extension\": \"false\", \"date_got_extension\": \"-1\", \"mask_id\": \"R7\", \"is_random_mask\": \"true\", \"created_by_api\": \"true\", \"has_website\": \"false\"}, \"timestamp\": 1709148528194}], \"ping_info\": {\"seq\": 0, \"start_time\": \"2024-02-28T19:28:48.194018+00:00\", \"end_time\": \"2024-02-28T19:28:48.194018+00:00\"}, \"client_info\": {\"telemetry_sdk_build\": \"glean_parser v12.0.1\", \"first_run_date\": \"Unknown\", \"os\": \"Unknown\", \"os_version\": \"Unknown\", \"architecture\": \"Unknown\", \"app_build\": \"Unknown\", \"app_display_version\": \"unknown\", \"app_channel\": \"local\"}}", "msg": "glean-server-event"}}
    {"Timestamp": 1709148528194563072, "Type": "request.summary", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.local", "EnvVersion": "2.0", "Severity": 6, "Pid": 87746, "Fields": {"errno": 0, "agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0", "lang": "en-US,en;q=0.5", "method": "POST", "path": "/api/v1/relayaddresses/", "uid": 3, "rid": "677474f3-172b-4e06-81f0-60ef2a558d8d", "t": 61}}
    ```
* Change the label of the mask
    - [ ] Among others, there should be a `glean-server-event` log, followed by a `request.summary` log, like:
    ```json
    {"Timestamp": 1709148608654010880, "Type": "glean-server-event", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.local", "EnvVersion": "2.0", "Severity": 6, "Pid": 87746, "Fields": {"document_namespace": "relay-backend", "document_type": "events", "document_version": "1", "document_id": "374e0998-ec59-46e9-a5f3-f26f7bdb3ece", "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0", "ip_address": "", "payload": "{\"metrics\": {}, \"events\": [{\"category\": \"email_mask\", \"name\": \"label_updated\", \"extra\": {\"client_id\": \"\", \"fxa_id\": \"fac5c63eb1c04dd4b9b0192f1a5483cb\", \"platform\": \"\", \"n_random_masks\": \"4\", \"n_domain_masks\": \"0\", \"n_deleted_random_masks\": \"1\", \"n_deleted_domain_masks\": \"0\", \"date_joined_relay\": \"1709058088\", \"premium_status\": \"email_unknown\", \"date_joined_premium\": \"-1\", \"has_extension\": \"false\", \"date_got_extension\": \"-1\", \"mask_id\": \"R7\", \"is_random_mask\": \"true\"}, \"timestamp\": 1709148608653}], \"ping_info\": {\"seq\": 0, \"start_time\": \"2024-02-28T19:30:08.653926+00:00\", \"end_time\": \"2024-02-28T19:30:08.653926+00:00\"}, \"client_info\": {\"telemetry_sdk_build\": \"glean_parser v12.0.1\", \"first_run_date\": \"Unknown\", \"os\": \"Unknown\", \"os_version\": \"Unknown\", \"architecture\": \"Unknown\", \"app_build\": \"Unknown\", \"app_display_version\": \"unknown\", \"app_channel\": \"local\"}}", "msg": "glean-server-event"}}
    {"Timestamp": 1709148608654702080, "Type": "request.summary", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.local", "EnvVersion": "2.0", "Severity": 6, "Pid": 87746, "Fields": {"errno": 0, "agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0", "lang": "en-US,en;q=0.5", "method": "PATCH", "path": "/api/v1/relayaddresses/7/", "uid": 3, "rid": "d041f4b3-376d-40a0-8188-96073ab894aa", "t": 50}}
    ```
 * In the profile menu, select "Manage your account" to go to Mozilla accounts. Under "Data Collection and Use", "Help improve ⁨Mozilla accounts⁩", turn from "On" to "Off". Return to Relay, sign out and sign in again. Update the label for a mask.
     - [ ] There is no `glean-server-event` log, just the `request.summary` log, like:
     ```json
     {"Timestamp": 1709148843867493888, "Type": "request.summary", "Logger": "fx-private-relay", "Hostname": "jwhitlock-MB-M2.local", "EnvVersion": "2.0", "Severity": 6, "Pid": 87746, "Fields": {"errno": 0, "agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:125.0) Gecko/20100101 Firefox/125.0", "lang": "en-US,en;q=0.5", "method": "GET", "path": "/api/v1/relayaddresses/", "uid": 3, "rid": "0abf6db1-7bd6-476f-a688-f73c9f38297e", "t": 5}}
     ```

groovecoder commented on privaterelay/glean_interface.py:
```diff
@@ -0,0 +1,326 @@
+    ) -> None:
+        """Log that a Relay email mask was created."""
+        user_data = UserData.from_user(mask.user)
+        if not user_data.metrics_enabled:
+            return
```
praise: woohoo honoring user data choices!

groovecoder commented on telemetry/glean/relay-server-metrics.yaml:
```diff
@@ -0,0 +1,159 @@
+    data_reviews:
+      - https://bugzilla.mozilla.org/show_bug.cgi?id=1882565
+    notification_emails:
+      - jwhitlock@mozilla.com
```
suggestion (non-blocking): `relay-team@mozilla.com` ?

jwhitlock replied:
The Glean documents have no guidance, so I reached out to internal Slack channel. According to @perrymcmanis144, we have to have a single person for metrics that do not expire. We could add a mailing list, but not replace the single person email.

However, `relay-team@mozilla.com` doesn't look like a good choice to me. There are many people on that list, including people no longer on the project. Once I have some experience with the kinds of emails sent to this address, we could add a maintained list with the people who should get notices and can do something about them.

groovecoder commented on telemetry/glean/relay-server-metrics.yaml:
```diff
@@ -0,0 +1,159 @@
+    data_sensitivity:
+      - interaction
+    extra_keys:
+      <<: *default_mask_extra_keys
```
praise: TIL yaml merge keys.

groovecoder commented on telemetry/glean/relay-server-metrics.yaml:
```diff
@@ -0,0 +1,159 @@
+        type: boolean
+  label_updated:
+    type: event
+    description: A Relay users updates an email mask's label.
```
```suggestion
    description: A Relay user updates an email mask's label.
```

groovecoder commented on privaterelay/tests/glean_tests.py:
```diff
@@ -0,0 +1,568 @@
+    """Check and return parts of the Glean payload that vary over test runs."""
+    event_ts_ms = payload["events"][0]["timestamp"]
+    event_time = datetime.fromtimestamp(event_ts_ms / 1000.0)
+    assert 0 < (datetime.now() - event_time).total_seconds() < 0.5
```
question (non-blocking): why do we assert that the event time must have been less than 0.5 seconds ago?

jwhitlock replied:
The timestamp should be in milliseconds. I'm trying to find a way to say "if we calculate the timestamp wrong, then it will be off by a factor of 1,000". This only tests the x1000 bigger, however...

groovecoder reviewed (changes requested):
Looked thru all the code; still need to test and spot-check. Biggest blocker is that we should NOT send a glean event when the email is blocked because of an error. We should only send a glean even when the email is blocked because of a user action.

jwhitlock commented on api/tests/views_tests.py:
```diff
@@ -89,8 +96,34 @@ def test_post_domainaddress_success(prem_api_client) -> None:
+    address = premium_user.domainaddress_set.get()
+    date_joined_ts = int(premium_user.date_joined.timestamp())
+    assert premium_user.profile.date_subscribed
+    date_subscribed_ts = int(premium_user.profile.date_subscribed.timestamp())
```
Added `user` argument to `create_expected_glean_event` to gather some of the items (`fxa_id`, `premium_status`, `date_joined_relay`, `date_joined_premium`) from the user.

jwhitlock commented on emails/models.py:
```diff
@@ -535,6 +535,38 @@ def metrics_enabled(self) -> bool:
             return bool(self.fxa.extra_data.get("metricsEnabled", True))
         return True
 
+    @property
+    def plan(self) -> Literal["free", "email", "phone", "bundle"]:
+        """The user's Relay plan as a string."""
+        if self.has_premium:
+            if self.has_phone:
+                return "bundle" if self.has_vpn else "phone"
+            else:
+                return "email"
+        else:
+            return "free"
```
This function is doing very little that would justify caching. If `has_premium` and `has_phone` are slow, should they be the cached properties? Or `fxa`?

groovecoder replied:
Possibly, but I think caching those caused some other stale data issues last time we tried it. I thought maybe caching at this level wouldn't cause those issues?

jwhitlock commented on privaterelay/glean_interface.py:
```diff
@@ -0,0 +1,326 @@
+        n_deleted_random_masks = user.profile.num_deleted_relay_addresses
+        n_deleted_domain_masks = user.profile.num_deleted_domain_addresses
+        date_joined_relay = user.date_joined
+        if user.profile.has_premium:
+            if user.profile.has_phone:
+                date_joined_premium = user.profile.date_subscribed_phone
+            else:
+                date_joined_premium = user.profile.date_subscribed
```
`date_joined_premium` is the timestamp for `premium_status`, and should change when the status changes. Our data doesn't track subscriptions this way, so this is a guess.

MPP-3764 is for a better solution

jwhitlock commented:
@groovecoder, this is ready for re-review. Changes since the last review:

* The data review in [bug 1882565](https://bugzilla.mozilla.org/show_bug.cgi?id=1882565) was approved. I'll return to that with the changes made since the review opened.
* I added the user to `create_expected_glean_event`, to centralize the boilerplate for checking user-derived fields
* I created an `AddressViewSet` base class to store the generalized `perform_create`, `perform_update`, and `perform_destroy` methods. I could not figure out a way to do this as a mixin that made `mypy` happy.
* Add comments for the leading 'R' and 'D' for `mask_id` metrics.
* Fixed a typo `users` → `user`
* Use a new function `log_email_dropped` to log details, with message `email_dropped`, for emails dropped for reasons other than mask blocking settings. Tests are updated to check for the new logs.
* Glean `email_blocked` events are now only for emails dropped due to mask blocking settings. Dropped `can_retry` from the event extra fields, since these can not be re-tried like some errors.
* Add some comments around turning Glean event time fields into Python `datetime`s.
* Add `# pragma: no cover` for some code that only runs when tests fail
* Fix a test that was exiting early instead of running all assertions.

jwhitlock commented on emails/tests/views_tests.py:
```diff
@@ -286,6 +303,34 @@ def _replace_mime_boundaries(email: str) -> str:
+            assert drop_log is None, "duplicate email_dropped log entry"
+            drop_log = record
+    assert drop_log is not None, "email_dropped log entry not found."
+    assert drop_log.levelno == logging.INFO
```
I only use `ERROR` logging if I want SREs to monitor and to get tracking Sentry issues. I didn't want that for these, but do you?

jwhitlock commented on emails/tests/views_tests.py:
```diff
@@ -1144,6 +1251,36 @@ def setUp(self) -> None:
+            event_time=timestamp,
+        )
+
+    def assert_my_log_email_dropped(
```
There's more than bounce tests, so I went with `assert_log_incoming_email_dropped`

jwhitlock commented on privaterelay/tests/glean_tests.py:
```diff
@@ -0,0 +1,565 @@
+        },
+    )
+    request_data = RequestData.from_request(request)
+    assert request_data.user_agent == "Mozilla/5.0 Firefox/125.0"
```
I added a test for just the user agent, and dropped the user agent from the IP tests.
